clientUtils/schemaProcessing.ts

```diff
@@ -0,0 +1,277 @@
+}
+
+function typeIsGivenOrNull(
+    typeToTest: string | string[],
+    type: string
+): boolean {
+    // We often have a type that is an array of a given type and null. This helper
+    // function returns true if the type is exactly the given type or an array of this and null
+    if (typeof typeToTest === "string") return typeToTest === type
+    return typeToTest.every((item) => item === type || item === "null")
+}
+
+function getEditorOptionForType(
+    type: string | string[],
+    enumOptions: string[] | undefined
+): EditorOption {
+    if (typeIsGivenOrNull(type, "number")) return EditorOption.numeric
+    else if (typeIsGivenOrNull(type, "string") && enumOptions)
+        return EditorOption.dropdown
+    else if (typeIsGivenOrNull(type, "string")) return EditorOption.textfield
+    else if (typeIsGivenOrNull(type, "boolean")) return EditorOption.checkbox
+    else if (typeIsGivenOrNull(type, "integer")) return EditorOption.numeric
+    else return EditorOption.textfield
+}
+
+/** This function takes a json schema that has been read and converted from json to a
+    JS object and processes it into a list of FieldDescriptions. A FieldDescription
+    contains just the information we will need for creating the columns for the Bulk FASTT
+    editor.
+
+    On a high level, what we do here is walk the recursive document structure of the json
+    schema and look at every level. If a property "type" exists and it is object then we
+    want to recurse over the properties (we are usually only interested in properties in
+    leaf position in the object graph because these are the ones we want to map to a single
+    spreadsheet cell). If the type is a primitive then we just create a FieldDescription entry.
+
+    There are some complications, for example:
+        * type can be an array to express a union of null | string
+        * anyof creates a fork that has to be combined into a single entry
+ *
+ */
+function extractSchemaRecursive(
+    schema: unknown,
+    pointer: string,
+    items: FieldDescription[]
+): void {
+    // We shouldn't encounter primitives in the schema itself (since we recurse
+    // only into expected structures which are almost always objects)
+    if (
+        isNull(schema) ||
+        isUndefined(schema) ||
+        isNumber(schema) ||
+        isString(schema) ||
+        isBoolean(schema) ||
+        isArray(schema)
+    ) {
+        console.error("shouldn't come here?", pointer)
+        return
+    } else if (isPlainObjectWithGuard(schema)) {
+        // If the current schema fragment describes a normal object
+        // then do not emit anything directly and recurse over the
+        // described properties
+        if (
+            schema.hasOwnProperty("type") &&
+            schema.type === "object" &&
+            schema.hasOwnProperty("properties")
+        ) {
+            for (const key of keys(schema.properties)) {
+                const newPath = `${pointer}/${key}`
```

Comment:
Is it safe to construct the path this way? Does it need to be escaped?

Reply:
The assumption here is that we control the schema so I think not - am I missing something?